keel-slack/src/main/kotlin/com/netflix/spinnaker/keel/slack/NotificationEventListener.kt

```diff
@@ -26,21 +32,26 @@ class NotificationEventListener(
   @EventListener(PinnedNotification::class)
   fun onPinnedNotification(notification: PinnedNotification) {
     with(notification) {
-      val envNotifications = repository.environmentNotifications(config.name, pin.targetEnvironment)
-      val deliveryArtifact = pin.reference.let {
-        repository.getArtifact(config.name, it)
-      }
+      val deliveryArtifact = config.artifacts.find {
+        it.reference == pin.reference
+      } ?: return
+
       val pinnedArtifact = repository.getArtifactVersion(deliveryArtifact, pin.version, null)
       val currentArtifact = repository.getArtifactVersionByPromotionStatus(config, pin.targetEnvironment, deliveryArtifact, PromotionStatus.CURRENT)
 
-      envNotifications.forEach { notificationConfig ->
-        if (notificationConfig.type == NotificationType.slack) {
-          val handler = handlers.find {
-            it.type == notification.type
-          } as? SlackNotificationHandler<SlackPinnedNotification>
-          handler?.let {
-            it.sendMessage(SlackPinnedNotification(
-              channel = notificationConfig.address,
+      val handler = handlers.find {
+        it.type == notification.type
+      } as? SlackNotificationHandler<SlackPinnedNotification> ?: return
+
+      if (pinnedArtifact == null || currentArtifact == null) {
+        log.warn("can't send notification as either pinned artifact or current artifacts information is missing")
+      } else {
+        config.environments.forEach { environment ->
```

Comment:
Why are we looping on every environment now? The pin is always for a specific environment, right?

Reply:
good call! I've refactored that piece of the code to get the notifications from the config, instead of making another call to the repository to fetch the notifications only. Will fix it!

Reply:
Sounds good. I made a suggestion for how to refactor this into a shared function that you can use for all notifications here: https://github.com/spinnaker/keel/pull/1751#discussion_r562822794

Reply:
done!